### ISSUE TITLE
Components become laggy

### ISSUE THREAD

**Issue by jimmyff** [dart-lang/angular_components#250]
After numerous components have been routed to and destroyed I am noticing that the angular_components become laggy. Some of the issues I've seen are:

- Tab triggers are delayed
- Tabs animation sometimes gets stuck half way through selecting another tab
- Material list triggers become delayed
- Material list item triggers on incorrect item
- Sometimes multiple taps are required to get an angular_component to trigger

This tends to only happen after a couple of minutes of using my app which suggests possibly a memory leak? (I'm aware I may have introduced this somehow). What's strange is the rest of the app remains fast, it's just the components become laggy.

You can see a screen recording here showing some of the issues listed above:
https://photos.app.goo.gl/UMjRFWybMRMTm8Kr2

I have a much more complex angular dart 4 app that doesn't experience the same issues. 

I can provide a link to see the web app if this would help.

**Comment by nshahan** [comment]
What versions of angular, angular_components, and the Dart SDK are you using?

Was that app built with dart2js or the dart-dev-compiler?

**Comment by jimmyff** [comment]
DDC, build runner and the latest versions of all the packages (as of 2 days ago). I'm away from laptop at the moment but will follow up with specific versions.

**Comment by jimmyff** [comment]
Hey @nshahan details below:

- angular: "5.0.0-alpha+10"
- angular_components: "0.9.0-alpha+10"
- Dart: 2.0.0-dev.47.0
- built with DDC

**Comment by nshahan** [comment]
Are you running this on a mobile device or a simulator with limited RAM? Have you tried a build with dart2js? The build with DDC is a debug build and expected to be much larger than a release build with dart2js. I have seen this cause memory issues in some instances when using the dev build in a environment with limited resources.

What compiler are you using for the build in the angular v4.x.x app?
I'd also recommend doing some memory profiling with chrome dev tools to see if you can pinpoint a leak and where it is coming from.

**Comment by jimmyff** [comment]
Oh apologies, I'm using dart2js for the build, forgot that i setup the `build.release.yaml` file: 
 
```
targets:
  $default:
    builders:
      build_web_compilers|entrypoint:
        generate_for:
          - web/main.dart
          - web/sw.dart
        options:
          compiler: dart2js
          dart2js_args:
            - --fast-startup
            - --minify
            - --trust-type-annotations
            - --trust-primitives
```
The capture is from a mobile device (Pixel 1). The Angular 4 app is built using `pub build` (dart2js)

I'll see if I can see anything with the chrome dev tools.

**Comment by jimmyff** [comment]
I've saved a chrome performance profile recorded on a pixel android device:

[pixel-profile-20180423T115208.zip](https://github.com/dart-lang/angular_components/files/1937804/pixel-profile-20180423T115208.zip)

It might mean more to you guys than me?  It was recorded from a fresh start and increasingly became more laggy during the recording.

**Comment by zacharydenton** [comment]
Based on your profile, I suggest:

1. Look into why your app is making so many calls to the Firebase API. There are 20+ Firebase requests, each taking 1s+ network time and 100ms+ CPU time.
2. Look into the size of the document. By the end of the profile, there are over 20000 nodes in the document, way too many for a mobile browser.

**Comment by jimmyff** [comment]
Thanks so much for the analysis. 

I knew I could optimise & cache the firebase db network requests but the nodes piling up is a surprise. The only thing I'm doing that's none-standard is using a proxy component to achieve two route paths on the same path (differentiating by navigation parameter string length). I added this as a shim whilst I wait for regex support in route paths ( https://github.com/dart-lang/angular/issues/1270 ). It's the only thing I can think that could explain the memory leak. I'll test to see if taking that out fixes the leak. 

Thanks again

**Comment by jimmyff** [comment]
I've taken out the proxy component but the app still gets laggy. Other than that it's a simple app, doesn't do anything fancy. Not sure why the nodes build up the way they do. I can share the project source with someone if that would help?

**Comment by nshahan** [comment]
The double-tap you are describing might be the same as https://github.com/dart-lang/angular_components_example/issues/66?

**Comment by jimmyff** [comment]
I have made some progress: I think the main part of the leak was due to a having my route definitions in the wrong order so the root route (`/`) was being created unnecessarily & unintentionally. 

The components do still become laggy after a while, although not nearly as quickly now. It now takes a couple of minutes of clicking through screens before the animations start skipping. 

I will do some more testing on my iOS test device and get back to you re: double tap.
